onnxruntime/core/providers/cuda/tensor/expand_impl.cu

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "expand_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T, int pad_mode>
+__global__ void _ExpandKernel(
+    const size_t rank,
+    const size_t N,
+    const size_t N_input,
+    const T* input_data,
+    T* output_data,
+    const int64_t* input_dims,
+    const int64_t* output_dims) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+  // initialize
+  int64_t output_index = id;
+  int64_t input_index = 0;
+
+  // use striding when tensor is larger than grid
+  int stride = blockDim.x * gridDim.x;
+  auto sizeTillDimensionOutput = N;
+  auto sizeTillDimensionInput = N_input;
+  auto out_coord = output_index;
+
+  // translate indices to coordinates. copy expanded dims from source
+  while (output_index < N) {
+    for (int64_t i = 0; i < rank; i++) {
+      sizeTillDimensionOutput /= output_dims[i];
+      sizeTillDimensionInput /= input_dims[i];
```

Comment:
Note that integer division is slow in CUDA. Please use fast_mod_div.

Reply:
Updated to use fast_divmod instead of integer division.

---
In reply to: [297309901](https://github.com/microsoft/onnxruntime/pull/1292#discussion_r297309901) [](ancestors = 297309901)